_posts/2020-06-19-async-explorer.markdown

```diff
@@ -0,0 +1,99 @@
+    warn("1) Hello there! Im taking a nap.\n", .{});
+    suspend;
+
+    warn("3) Why did you wake me up? Im going back to sleep.\n", .{});
+    suspend;
+
+    warn("5) Fine, here you go.\n", .{});
+    return x+y;
+}
+
+fn amain() void {
+    var frame = async foo(1, 2);
+    
+    warn("2) Wake up!\n", .{});
+    resume frame;
+
+    warn("4) Grab a brush and put a little (makeup)!\n", .{});
+    resume frame;
+
+    var ret = await frame;
+    warn("6) foo(1,2) == {}\n", .{ret});
+}
+```
+
+Let's review the flow of this program. We call `foo()` using the `async` keyword, and at that point, `foo()` begins executing. Afterwards, using the keyword `suspend`, `foo()` forfeits it's context, which gives `amain()` an opaque `frame`, which it can use to "wake up" `foo()`. As we can see `foo()` forfeits it's context again, and finally, it returns a return value, using the `return` keyword. At that point, `amain()` retrieves that value using the `await` keyword.
+
+So the expected output is something like this:
+
+```text
+D:\Projects\AsyncExplorer\examples> .\simple_async.exe
+1) Hello there! Im taking a nap.
+2) Wake up!
+3) Why did you wake me up? Im going back to sleep.
+4) Grab a brush and put a little (makeup)!
+5) Fine, here you go.
+6) foo(1,2) == 3
+```
+
+Phew. That was exhausting. But I think we have a pretty solid grip on what the basics of "async" means.
+
+## Examples
+
+## Use cases
+
+Now I hope we understand the _definition_ of a coroutine, but what are the uses of these lazy little gremlins? Are they in some way _stronger_ than our classical subroutines?
+
+The short answer is _No_. Coroutines (and more generally, every "code construct" of that sort) doesn't allow us to calculate _more things_. But we are developers, not computer scientists. What we care about is having extensible, modular, well designed solutions to our problems. That means, we like _modelling_ our problems in such ways that let us interact with them conviniently through code. For example through Object-Oriented design, Design Patterns, and in our cases, coroutines.
+
+Im not going to go into the use cases of coroutines too much, since it's not the point of this series. I will, though, give a quick run down of what our "break taking" functions can be usful for:
+
+* Modelling [Generators](https://en.wikipedia.org/wiki/Generator_(computer_programming)).
+* Implementing [Cooperative Multitasking](https://en.wikipedia.org/wiki/Cooperative_multitasking).
+* Using [Asynchronous IO](https://en.wikipedia.org/wiki/Asynchronous_I/O).
+* A lot of other cool stuff.
+
+## Conclusion
+
+The point of this post was to "set the stage" for a couple of posts I'll publish here in the future. I hope I managed to give you the idea of "what" coroutines are, and maybe a touch of their use-case. In the next posts we'll be digging into some of that oh-so-sweet x86 assembly, taking a look of some the _implementations_ of coroutines provided by some compiled languages.
```

Comment:
Add link to "x86 assembly", or mark it as a term `like that`

Reply:
good idea